contracts/DKVDaggerHashimoto.sol

```diff
@@ -223,6 +232,45 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
         return hash0;
     }
 
+    /*
+     * Run a modified hashimoto hash,
+     * with Merkle inclusion proofs
+     */
+    function _hashimotoMerkleProof(
+        uint256 startShardId,
+        uint256 shardLenBits,
+        bytes32 hash0,
+        bytes32[][] memory proofsDim2,
+        bytes[] memory maskedData
+    ) internal view returns (bytes32) {
+        require(maskedData.length == randomChecks, "data vs checks: length mismatch");
+        require(proofsDim2.length == randomChecks, "proofs vs checks: length mismatch");
+        uint256 maxKvSize = 1 << maxKvSizeBits;
+        uint256 rows = 1 << (shardEntryBits + shardLenBits + chunkLenBits);
+
+        for (uint256 i = 0; i < randomChecks; i++) {
+            require(maskedData[i].length <= chunkSize, "invalid proof size");
+            uint256 parent = uint256(hash0) % rows;
+            uint256 chunkIdx = parent + (startShardId << (shardEntryBits + chunkLenBits));
+            uint256 kvIdx = chunkIdx >> chunkLenBits;
+            chunkIdx = chunkIdx % (1 << chunkLenBits);
+            bytes memory data = maskedData[i];
+            /* NOTICE: Now we have kvIdx and chunkIdx both generated from hash0
+             *         The difficulty should increase intrinsically */
+            require(
+                systemContract.checkDaggerDataWithProof(chunkIdx, kvMap[idxMap[kvIdx]].hash, proofsDim2[i], data),
```

Comment:
We are changing the API to `systemContract.checkDaggerData` to `systemContract.unmaskData`, which will return unmasked data so that the inclusion verification check can be done in this contract.

PR is here https://github.com/QuarkChain/go-ethereum/pull/110

Reply:
Can we make it as a separate pr? Considering it not quite related to this topic, I believe it is an orthogonal effort.
We might change all involved functions call and tests.